server/api/utils/cache.py

```diff
@@ -0,0 +1,131 @@
+# Copyright 2023 Iguazio
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+import asyncio
+import contextlib
+import typing
+import uuid
+
+
+class CachedObject:
+    """
+    Base class for objects that are cached in the Cache class.
+    """
+
+    def __init__(
+        self,
+        object_id: str,
+        lock: asyncio.Lock = None,
+        expiry_delayed_call: asyncio.Handle = None,
+    ):
+        self._object_id = object_id
+        self.lock = lock or asyncio.Lock()
+        self._expiry_delayed_call: typing.Optional[asyncio.Handle] = expiry_delayed_call
+
+    def matches(self, object_id: str) -> bool:
+        return self._object_id == object_id
+
+    def __exit__(self, exc_type, exc_val, exc_tb):
+        if self._expiry_delayed_call:
+            self._expiry_delayed_call.cancel()
+
+
+class Cache:
+    """
+    A cache for objects that are created asynchronously and need to be locked while in use.
+    """
+
+    def __init__(self, cls):
+        self.lock = asyncio.Lock()
+        self.cls = cls
+        self.cache: typing.Dict[str, CachedObject] = {}
+
+    @contextlib.asynccontextmanager
+    async def get_or_create_locked(
+        self,
+        key: str,
+        ttl: float,
+        cls_kwargs: dict = None,
+    ) -> typing.Tuple[CachedObject, bool]:
+        """
+        Get an object from the cache, or create it if it doesn't exist.
+        The object will be locked while the context is active.
+        :param key: The object's key in the cache.
+        :param ttl: The object's time to live in the cache.
+        :param cls_kwargs: Keyword arguments to pass to the object's constructor.
+
+        :return:    A CachedObject instance and a boolean indicating whether the object was created.
+        """
+        created = False
+        async with self.lock:
+            obj = self.cache.get(key)
+            if not obj:
+                created = True
+                obj = self._create(key, ttl, cls_kwargs)
+
+            # lock the run specific lock before releasing the global lock
+            await obj.lock.acquire()
```

Comment:
This has deadlock potential - 
If someone else locked the object and trying to do something requiring the global lock (insert a new object), this context will be stuck on the object lock, while the other context will be stuck waiting for us to release the global lock.
If you keep this order of locks, then it needs to be at least documented or in some other way prevented that the global lock is acquired when an object lock is held.

Reply:
nice catch, will rethink it